lib/pact_broker/index/service.rb

```diff
@@ -194,6 +184,39 @@ def self.latest_verifications_for_consumer_version_tags(options)
           nil # should not be used
         end
       end
+
+      def self.base_query(options)
+        query = PactBroker::Pacts::PactPublication.select(Sequel[:pact_publications][:id])
+
+        if options[:consumer_name]
+          consumer = pacticipant_repository.find_by_name!(options[:consumer_name])
+          query = query.for_consumer(consumer)
+        end
+
+        if options[:provider_name]
+          provider = pacticipant_repository.find_by_name!(options[:provider_name])
+          query = query.for_provider(provider)
+        end
+
+        query
+      end
+
+      def self.pact_publication_ids_query(query, tags)
+        return query.union(query.latest_for_consumer_tag(tags)) if tags.is_a?(Array)
+        return query.union(query.latest_by_consumer_tag) if tags
+
+        query
+      end
+
+      def self.pacticipant_name_query(pacticipant_name)
```

Comment:
This is getting closer! But it's still filtering a really big slow dataset (the entire index query) on a string match, rather than finding the pacticipants (a really small query) on a string match, and then being able to set the exact IDs of the consumer and provider in the big query (which will make it super quick). What's your thinking behind that?